server/src/main/java/org/elasticsearch/search/retriever/KnnRetrieverBuilder.java

```diff
@@ -111,20 +115,64 @@ public KnnRetrieverBuilder(
         Float similarity
     ) {
         this.field = field;
-        this.queryVector = queryVector;
+        this.queryVector = () -> queryVector;
         this.queryVectorBuilder = queryVectorBuilder;
         this.k = k;
         this.numCands = numCands;
         this.similarity = similarity;
     }
 
-    // ---- FOR TESTING XCONTENT PARSING ----
+    private KnnRetrieverBuilder(KnnRetrieverBuilder clone, Supplier<float[]> queryVector, QueryVectorBuilder queryVectorBuilder) {
+        this.queryVector = queryVector;
+        this.queryVectorBuilder = queryVectorBuilder;
+        this.field = clone.field;
+        this.k = clone.k;
+        this.numCands = clone.numCands;
+        this.similarity = clone.similarity;
+        this.retrieverName = clone.retrieverName;
+        this.preFilterQueryBuilders = clone.preFilterQueryBuilders;
+    }
 
     @Override
     public String getName() {
         return NAME;
     }
 
+    @Override
+    public RetrieverBuilder rewrite(QueryRewriteContext ctx) throws IOException {
+        var rewrittenFilters = rewritePreFilters(ctx);
+        if (rewrittenFilters != preFilterQueryBuilders) {
+            var rewritten = new KnnRetrieverBuilder(this, queryVector, queryVectorBuilder);
+            rewritten.preFilterQueryBuilders = rewrittenFilters;
+            return rewritten;
+        }
+
+        if (queryVectorBuilder != null) {
+            SetOnce<float[]> toSet = new SetOnce<>();
+            ctx.registerAsyncAction((c, l) -> {
+                queryVectorBuilder.buildVector(c, l.delegateFailureAndWrap((ll, v) -> {
+                    toSet.set(v);
+                    if (v == null) {
+                        ll.onFailure(
+                            new IllegalArgumentException(
+                                format(
+                                    "[%s] with name [%s] returned null query_vector",
+                                    QUERY_VECTOR_BUILDER_FIELD.getPreferredName(),
+                                    queryVectorBuilder.getWriteableName()
+                                )
+                            )
+                        );
+                        return;
+                    }
+                    ll.onResponse(null);
+                }));
+            });
+            var rewritten = new KnnRetrieverBuilder(this, () -> toSet.get(), null);
+            return rewritten;
+        }
+        return super.rewrite(ctx);
```

Comment:
Should we make the parent `RetrieverBuilder#rewrite` abstract to ensure that all retrievers have to implement rewrite?